api/api.py

```diff
@@ -766,6 +769,9 @@ def buy_listing(user, key):
     if listing.available_quantity < quantity:
         return jsonify({'message': "Not enough items in stock!"}), 400
 
+    # NB: here we "lock" the quantity. it is given back if the sale expires
+    listing.available_quantity -= quantity
```

Comment:
I think we should put this at the bottom, just before creating the Sale. That way if we get any of the exceptions (`kraken api` is down, `get_new_address` doesn't work...), we don't remove the quantity that will not return anyways because the sale did not get created.

You could even put it just before `return jsonify({'sale': sale.to_dict()})`, because there you know that the Sale got created and saved into the database. Or is the quantity also saved to the database when `db.session.commit()`?

Reply:
Yes, `.commit()` does the whole thing. :) Before that, everything is temporary, including the inflation in the quantity of items....